### DISCUSSION TITLE
Best practice from Rust's HashMap to Java Map?

### DISCUSSION THREAD

**Discussion by tisonkun** [jni-rs/jni-rs#495] [Q&A] [answered]
A `java.util.HashMap` should be fine. But currently it's a bit complicated to wrap a utility for doing so, compared with `env.new_string` for converting strings.

**Comment by tisonkun** [comment]
I wrote a utility for OpenDAL's usage:

https://github.com/apache/incubator-opendal/pull/3069

```rust
fn hashmap_to_jmap<'a>(env: &mut JNIEnv<'a>, map: &HashMap<String, String>) -> Result<JObject<'a>> {
    let map_object = env.new_object("java/util/HashMap", "()V", &[])?;
    let jmap = env.get_map(&map_object)?;
    for (k, v) in map {
        let key = env.new_string(k)?;
        let value = env.new_string(v)?;
        jmap.put(env, &key, &value)?;
    }
    Ok(map_object)
}
```

**Comment by argv-minus-one** [comment, marked as answer]
`JNIEnv::new_string` is a wrapper around a C JNI function; it's built into the JVM.

Converting between Java and Rust types in general is another story altogether. The way you've done it pretty much is the best practice at this point. Perhaps someone will write some conversion traits and/or a JNI binding generator in the future, but that would be a lot of work.

**Reply by tisonkun** [reply]
Make sense.
